### PR TITLE
Class to read and process GPS readings (with ublox F9R rover GNSS module)

### LINKED ISSUE #161: Acquisition and parsing of GPS data
### Describe the solution you'd like

I have uploaded the data we collected on 28/11/23 [https://gist.github.com/alphaville/995a15ef74e8d5b3128c87667b5bfdaf](here). This is how the data looks like

```
$GPGLL,5434.81291,N,00556.15593,W,145737.00,A,A*70
$GPRMC,145738.00,A,5434.81285,N,00556.15627,W,0.351,,281123,,,A*63
$GPVTG,,T,,M,0.351,N,0.650,K,A*27
$GPGGA,145738.00,5434.81285,N,00556.15627,W,1,05,1.28,64.4,M,52.5,M,,*79
$GPGSA,A,3,12,32,24,15,19,,,,,,,,2.51,1.28,2.16*02
$GPGSV,3,1,10,02,02,344,37,12,56,202,22,15,26,166,27,17,,,35*49
$GPGSV,3,2,10,19,34,071,36,22,11,050,32,23,18,236,,24,72,108,35*72
$GPGSV,3,3,10,25,21,223,,32,26,309,41*75
```

### Documentation

The data are structured according to the [NMEA Standard Logs](https://docs.novatel.com/OEM7/Content/Logs/NMEA_Standard_Logs_.htm) protocol. Examples:

1. [GPGLL](https://docs.novatel.com/OEM7/Content/Logs/GPGLL.htm) - Geographic Position
2. [GPGSV](https://docs.novatel.com/OEM7/Content/Logs/GPGSV.htm) - GPS Satellites in View
3. [GPVTG](https://docs.novatel.com/OEM7/Content/Logs/GPVTG.htm) - Track Made Good and Ground Speed

First of all we need to decipher what all these mean and how the data are structured (read the docs).

### Issues to address

1. We interface the GNSS over the serial at a baud rate of `500,000`. We can write a simple interface that reads the data from serial (using `pyserial`) and parses it
2. We should follow what we did [here](https://github.com/QUB-ASL/bzzz/blob/dev/raspberry/bzzz/sensors/anemometer.py) for the anemometer: a thread runs in the background, collects data and stores it in memory, and then we can access is with a non-blocking function (or get a moving average, median, etc). 
3. For some reason the data seems to be coming in at 1Hz, which is too slow. Can we speed this up?

### PR DESCRIPTION
**Pull request by alphaville**
## Main Changes

- New classes for parsing and handling GPS data

## Associated Issues

- Closes #161

## Tests

⚠️ Needs to be tested! ⚠️

### CODE REVIEW

**Comment by alphaville** [comment]
@pdavid747 How is this progressing? Let us know if you need help.

**Comment by pdavid747** [comment]
@alphaville today I got the base station to provide correction data to the rover, I will be testing it hopefully tomorrow outside. I will finish this after that is done so I can move on to implement a KF.

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 1]
```diff
@@ -11,46 +10,59 @@ def deg_min_sec_to_decimal(
     degrees, minutes, direction
     ):
     """
-    Convert degrees-minutes-seconds from GPGLL to degrees (deciman)
+    Convert degrees-minutes-seconds from GNGLL to degrees (deciman)
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps_lea6s.py:13]: decimal?

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 2]
```diff
@@ -11,46 +10,59 @@ def deg_min_sec_to_decimal(
...
 
     :param dms: degrees-minutes-seconds (must be positive)
     :param dir: direction (can be one of 'N', 'W', 'E', 'S')
     """
     
     decimal_degrees = degrees + minutes / 60
-    if direction in ['S', 'W']:
+    if direction in ['S', 'W']:  # South and West lat and long need to be denoted with a "-" value
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps_lea6s.py:20]: "need to be denoted with a "-" value" -> are negative

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 3]
```diff
@@ -11,46 +10,59 @@ def deg_min_sec_to_decimal(
...
-
+    
+    """
+    Initializes the GPS data handler with default serial path and baud rate, 
+    sets up storage for latitude, longitude, and altitude, 
+    and starts a background thread to continuously read and parse GPS data.
+    """
     def __init__(self,
-                 serial_path="/dev/tty.usbmodem142201",
+                 serial_path="/dev/tty.usb...",
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps_lea6s.py:32]: Does `"/dev/tty.usb..."` work?

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 4]
```diff
@@ -89,10 +98,19 @@ def longitude(self):
...
+    
+    @property
+    def altitude(self):
+        """
+        GPS data object
+        """
+        return self.__gpgsv_altitude
 
-my_sensor = GpsLea6SReader(serial_path="/dev/tty.usbmodem142201")
+my_sensor = GpsLea6SReader(serial_path="/dev/tty.usb...")
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps_lea6s.py:109]: Call the sensor using a context - i.e., 

```python
with GpsLea6SReader(serial_path="/dev/tty.usb...") as mysensor:
   # your code goes here
```

Comment (pdavid747) [reply raspberry/bzzz/sensors/gps_lea6s.py:109]: the wrong version of the code has been committed I will fix now

Comment (pdavid747) [reply raspberry/bzzz/sensors/gps_lea6s.py:109]: very strange, it says everything has been committed on my branch, but it has not on git hub for some reason...

Comment (pdavid747) [reply raspberry/bzzz/sensors/gps_lea6s.py]: all fixed now

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    :param dms: degrees-minutes-seconds (must be positive)
+    :param dir: direction (can be one of 'N', 'W', 'E', 'S')
+    """
+    
+    decimal_degrees = degrees + minutes / 60
+    if direction in ['S', 'W']:  # need to be denoted with a "-" value -> are negative
+        decimal_degrees *= -1
+    return decimal_degrees
+
+class GpsLea6SReader:
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:24]: should we change this to ZED-F9R instead of LEA6S

Comment (alphaville) [reply raspberry/bzzz/sensors/gps_lea6s.py:24]: This code is an NMEA parser. It's independent of the module. We should change it to something more generic.

**Comment by alphaville** [comment]
@pdavid747 has there been any progress on this? Can you push your changes so we can review this PR again?

**Comment by pdavid747** [comment]
@alphaville changes have been pushed

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
+from math import degrees
+import serial
+import time
+from threading import Thread
+import numpy as np
+
+
+def deg_min_sec_to_decimal(
+    #dms, dir="N"
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:9]: Think we can remove commented out Param's

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    degrees, minutes, direction
+    ):
+    """
+    Convert degrees-minutes-seconds from GNGLL to decimal
+
+    :param dms: degrees-minutes-seconds (must be positive)
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:15]: These param need updated

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    """
+    decimal_degrees = degrees + minutes / 60
+    if direction in ['S', 'W']:  # need to be denoted with a negative value
+        decimal_degrees *= -1
+    return decimal_degrees
+
+class GpsReader:
+    
+    """
+    Initializes the GPS data handler with default serial path and baud rate,
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:26]: Initialises

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                    self.__gpgsv_altitude = altitude
+                        
+                except Exception as e:
+                    print(f"Error processing GPS data: {e}")
+
+
+    @property
+    def latitude(self):
+        """
+        GPS data object
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:90]: Change the description to what the function does
Returns the latitude ...

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        """
+        return self.__gpgll_latitude
+
+    @property
+    def longitude(self):
+        """
+        GPS data object
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:97]: Same here

File: raspberry/bzzz/sensors/gps_lea6s.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        """
+        return self.__gpgll_longitude
+    
+    @property
+    def altitude(self):
+        """
+        GPS data object
```
Comment (jamie-54) [inline raspberry/bzzz/sensors/gps_lea6s.py:104]: and here

**Comment by jamie-54** [comment]
I've made a few suggestions, maybe change the name of the script to something more suitable as well

**Comment by alphaville** [comment]
@pdavid747 I renamed the file to `gps.py` and the class to `Gps`. 

@ejb-11 @Yuanbwcx @Minanchi @Runway27 we are still waiting for your reviews.

File: docs/whitepapers/kalmanfilter/kalmanfilter.bbl
Comment (alphaville) [inline docs/whitepapers/kalmanfilter/kalmanfilter.bbl:1]: @pdavid747 It seems that you have pushed some irrelevant files to this PR

File: raspberry/bzzz/sensors/gps.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        """
+        return self.__gpgsv_altitude
+
+if __name__ == "__main__":
+    gps = Gps()  # Using default serial path and baud rate
+    time.sleep(120)  # Keep the main program running for 2 minutes
+
+
+
+# from math import degrees
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps.py:109]: We shouldn't have commented-out code here

File: raspberry/bzzz/sensors/gps.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+
+    @property
+    def altitude(self):
+        """
+        Altitude in meters
+        """
+        return self.__gpgsv_altitude
+
+if __name__ == "__main__":
+    gps = Gps()  # Using default serial path and baud rate
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps.py:104]: Use the syntax 

```python
with Gps() as gps:
    time.sleep(120)
```

For this to work you need to implement `__enter__` and `__exit__`. Can you see how we did this for the [anemometer](https://github.com/QUB-ASL/bzzz/blob/dev/raspberry/bzzz/sensors/anemometer.py) and do the same here?

File: raspberry/bzzz/sensors/gps_data.csv
[the remark this answers is not in the document]
Comment (pdavid747) [reply raspberry/bzzz/sensors/gps_data.csv:1]: no problem, I just included it so everyone could see data from the test

File: raspberry/bzzz/sensors/gps.py [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+    :param minutes: minutes
+    :param direction: direction (can be one of 'N', 'W', 'E', 'S')
+    :return: decimal representation of GPS coordinates
+    """
+    decimal_degrees = degrees + minutes / 60
+    if direction in ['S', 'W']:  # need to be denoted with a negative value
+        decimal_degrees *= -1
+    return decimal_degrees
+
+class GpsLogger:
```
Comment (alphaville) [inline raspberry/bzzz/sensors/gps.py:21]: Why don't you use the [other logger](https://github.com/QUB-ASL/bzzz/blob/dev/raspberry/bzzz/sensors/data_logger.py) we already have?

**Pull request closed without merging**